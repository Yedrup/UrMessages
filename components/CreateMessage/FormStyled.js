import styled, { keyframes } from 'styled-components';

export const loading = keyframes`
  from {
    background-position: 0 0;
  }
  to {
    background-position: 100% 100%;
  }
`;

const FormStyled = styled.form`
  max-width: 80rem;
  margin: 2rem auto;
  border: 2px solid grey;
  padding: 1.5rem;
  line-height: 1.5;
  font-size: 1.5rem;
  font-weight: 600;
  label {
    display: block;
    margin-bottom: 1rem;
    text-align: left;
  }
  input,
  textarea,
  select {
    width: 100%;
    padding: 0.5rem;
    font-size: 1.5rem;
    border: 1px solid black;
    background: white;
    &:focus {
      outline: 0;
      border-color: ${({ theme }) => theme.highlight};
    }
  }
  textarea {
    min-height: 7rem;
  }
  button {
    align-self: center;
  }
  button,
  input[type='submit'] {
    font-size: 1.85rem;
    font-weight: 600;
    text-transform: uppercase;
  }
  fieldset {
    border: 0;
    padding: 0;

    &[disabled] {
      opacity: 0.45;
    }
    &::before {
      height: 10px;
      content: '';
      display: block;
      background-image: linear-gradient(
        to right,
        #a14ae2 0%,
        ${({ theme }) => theme.highlight} 50%,
        #a14ae2 100%
      );
    }
    &[aria-busy='true']::before {
      background-size: 50% auto;
      animation: ${loading} 5s linear infinite;
    }
  }
`;

export default FormStyled;
