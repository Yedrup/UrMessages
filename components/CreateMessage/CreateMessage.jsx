import { useContext, useState, useEffect } from 'react';
import styled from 'styled-components';
import NProgress from 'nprogress';
import { DataDispatchContext } from '../../contexts/DataContext';
import { UIDispatchContext, UIStateContext } from '../../contexts/UIContext';
import { UserStateContext } from '../../contexts/UserContext';
import { MessageClass } from '../../lib/classes';
import { postMessage } from '../../lib/data-treatment-service';
import SigninButton from '../Sign/SigninButton';
import FormStyled from './FormStyled';
import Error from '../UIFeedbackMessages/ErrorMessage';
import Info from '../UIFeedbackMessages/InfoMessage';

const InfoStyled = styled(Info)`
  display: flex;
  align-items: center;
  button {
    margin-left: 2rem;
    padding: 1rem 0;
    background: ${({ theme }) => theme.highlightSecondary};
    color: white;
    font-size: 1.5rem;
  }
`;

const CreateMessage = ({ ctx }) => {
  const { id, isPublic, parentId } = ctx;
  const type = isPublic ? 'public' : 'private';

  const { dispatchData } = useContext(DataDispatchContext);
  const { dispatchUI } = useContext(UIDispatchContext);
  const { stateUI } = useContext(UIStateContext);
  const { stateUser } = useContext(UserStateContext);

  let defaultFormValues = { title: '', content: '', type };
  const [values, setValues] = useState(defaultFormValues);
  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const [error, setError] = useState(null);

  // Reset the form on disconnection
  useEffect(() => {
    if (!stateUser.isConnected) setValues({ ...defaultFormValues });
    return () => {};
  }, [stateUser?.isConnected]);

  const handleChange = e => {
    e.preventDefault();
    const { name, value } = e.target;
    setValues({ ...values, [name]: value });
  };

  const sendMessage = async ({ title, content, type }) => {
    dispatchUI({
      type: 'START_BUSY',
    });
    setIsFormSubmitting(true);
    NProgress.start();
    let newFormattedMessage = new MessageClass({
      id,
      parentId,
      isPublic,
      name,
      title,
      user: stateUser?.userId,
      content,
    });
    await postMessage({ message: newFormattedMessage })
      .then(result => {
        // Dispatch the new list updated
        dispatchData({
          type: 'CREATE_MESSAGE',
          payload: { result, type },
        });
        dispatchUI({
          type: 'END_BUSY',
        });
        // Reset the form on success
        setValues({ ...defaultFormValues });
      })
      .catch(err => {
        setError(err);
        dispatchUI({
          type: 'IS_ERROR',
          payload: { ...err },
        });
        dispatchUI({
          type: 'END_BUSY',
        });
      });
    NProgress.done();
    setIsFormSubmitting(false);
  };

  return (
    <>
      {!stateUser?.isConnected && (
        <InfoStyled info={{ message: 'You need to Sign In to send messages!' }}>
          <span className="createMessage__info--signin">
            <SigninButton />
          </span>
        </InfoStyled>
      )}
      <FormStyled
        data-test="form"
        id="test"
        onSubmit={async e => {
          e.preventDefault();
          sendMessage(values);
        }}
      >
        <fieldset
          disabled={stateUI?.isBusy || !stateUser.isConnected}
          aria-busy={stateUI?.isBusy}
        >
          {error && stateUI.isError && <Error error={error} />}

          <label htmlFor="title">
            Title
            <input
              type="text"
              id="title"
              name="title"
              placeholder="Title"
              required
              onChange={handleChange}
              value={values.title}
            />
          </label>
          <label htmlFor="content">
            Message
            <textarea
              id="content"
              name="content"
              placeholder="Enter A Message"
              required
              onChange={handleChange}
              value={values.content}
            />
          </label>
          <button
            type="submit"
            className="c-button"
            data-testid="button-submit-message"
          >
            Submit{stateUI?.isBusy && isFormSubmitting ? 'ting' : ''}
          </button>
        </fieldset>
      </FormStyled>
    </>
  );
};

export default CreateMessage;
