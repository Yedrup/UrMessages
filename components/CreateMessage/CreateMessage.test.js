import { render, cleanup } from '@testing-library/react';
import PropTypes from 'prop-types';
import DataProvider from '../../contexts/DataContext';
import UIProvider from '../../contexts/UIContext';

import CreateMessage from './CreateMessage';

const ContextWrapper = ({ children }) => (
  <DataProvider>
    <UIProvider>{children}</UIProvider>
  </DataProvider>
);

ContextWrapper.propTypes = {
  children: PropTypes.object,
};

const fakeCtx = {
  id: '1234',
  isPublic: false,
  parentId: '9876',
};

describe('<CreateMessage/>', () => {
  afterEach(cleanup);

  it('properly renders the component - snapshot', () => {
    const { asFragment } = render(
      <ContextWrapper>
        <CreateMessage ctx={fakeCtx} />
      </ContextWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
