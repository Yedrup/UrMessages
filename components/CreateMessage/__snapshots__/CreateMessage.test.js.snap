// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateMessage/> properly renders the component - snapshot 1`] = `
<DocumentFragment>
  <form
    class="FormStyled-j8ogoe-0 iIBCUw"
    data-test="form"
    id="test"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="title"
      >
        Title
        <input
          id="title"
          name="title"
          placeholder="Title"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="content"
      >
        Message
        <textarea
          id="content"
          name="content"
          placeholder="Enter A Message"
          required=""
        />
      </label>
      <button
        class="c-button"
        data-testid="button-submit-message"
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>
</DocumentFragment>
`;
